### PR TITLE
Lido app/marcus/json

### PR DESCRIPTION
**Pull request by marhong2**
Had to create new PR, b/c recent pushes to master were showing.

### CODE REVIEW

**Comment by marhong2** [comment]
Code for converting to JSON:


const tests = require('./TESTS.json');
    saveTests: (TestCase, type) => {
        const json = (tests == null)? {} : tests,
              functions = {},
              funs = Object.keys(TestCase);
        funs.forEach(fun => {
            const formulaValuePairs = {};
            fun.forEach(formula => {
                formulaValuePairs[formula] = TestCase[fun][formula]
            })
            functions[fun] = formulaValuePairs
        })
        json[type] = formulaValuePairs;

        fs.writeFile(file, JSON.stringify(json), function (err) {
            if (err) throw err;
        });
    }

**Comment by marhong2** [comment]
Note: Yarn.lock changed. I am assuming b/c I imported Sinon to fake out the system clock

File: test/formulas/allSheets.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
+const FormulaError = require("../../formulas/error")
+
+FormulaError
+module.exports = {
```
Comment (tyleradams) [inline test/formulas/allSheets.js:4]: Put this in the JSON for each test. Each test should have a "sheets" param.

File: test/utils.js [hunk 2]
```diff
@@ -1,26 +1,46 @@
...
-                const formulas = Object.keys(TestCase[fun]);
+                const formulas = Object.keys(TestCase[fun])
                 formulas.forEach(formula => {
-                    const expected = TestCase[fun][formula];
-                    let result = parser.parse(formula, {row: 1, col: 1});
+                    let expected = TestCase[fun][formula]
+                    let result = parser.parse(formula, {row:1, col:1});
                     if (result.result) result = result.result;
+                    //variable in nature, so I can not save them in a JSON
+                    if(formula === "HOUR(NOW())")
```
Comment (tyleradams) [inline test/utils.js:23]: Override the date function so it's deterministic.
https://stackoverflow.com/questions/1976968/change-browser-time-to-test-return-value-of-date

File: test/utils.js [hunk 3]
```diff
@@ -1,26 +1,46 @@
 const expect = require('chai').expect;
+const fs = require('fs')
+const sinon = require('sinon')
 module.exports = {
-    generateTests: (parser, TestCase) => {
-        const funs = Object.keys(TestCase);
-
+    generateTests: (parser, TestCase, type) => {
+        let clock = null;
+            now = new Date();
```
Comment (tyleradams) [inline test/utils.js:7]: Use 2020-01-01T00:00:00Z as now so it's determinstic, it's the whole reason why we're mocking out the clock

File: formulas/error.js [hunk 4]
```diff
@@ -94,6 +94,26 @@ FormulaError.REF = new FormulaError("#REF!");
  */
 FormulaError.VALUE = new FormulaError("#VALUE!");
 
+FormulaError.stringToError = (str) => {
+  switch(str) {
```
Comment (tyleradams) [inline formulas/error.js:98]: Use a map.

File: test/formulas/master.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+const FormulaError = require('../../formulas/error');
+const {FormulaParser} = require('../../grammar/hooks');
+const {generateTests} = require('../utils');
+const json = require('./TESTS.json')
+const test = function(testName){
+    let data = json[testName]["SHEET"]
+    data.map((row, rowIndex) => row.map((value, colIndex) => {
```
Comment (tyleradams) [inline test/formulas/master.js:7]: If you don't use the result, use a forEach.

File: test/formulas/master.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        if(value === "undefined")
+            data[rowIndex][colIndex] = undefined
+        if(FormulaError.stringToError(value))
+            data[rowIndex][colIndex] = FormulaError.stringToError(value);
+    }));
+    const parser = new FormulaParser({
+        //used in all tests except for text,
+        onCell: ref => {
+            return data[ref.row - 1][ref.col - 1]
```
Comment (tyleradams) [inline test/formulas/master.js:16]: 1 index the data to make the code simpler.

File: test/formulas/master.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        },
+        //used in date, engineering, logical,
```
Comment (tyleradams) [inline test/formulas/master.js:18]: Irrelevant and subject to change

File: test/formulas/master.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                const innerArr = [];
+                for (let col = ref.from.col - 1; col < ref.to.col; col++) {
+                    innerArr.push(data[row][col])
+                }
+                arr.push(innerArr);
+            }
+            return arr;
+        },
+        //only used with trigonometry
+        onVariable: name => {
```
Comment (tyleradams) [inline test/formulas/master.js:33]: Fold variables into the json file per test.

**Review by tyleradams — CHANGES_REQUESTED** [review]
Looking good!

File: formulas/error.js [hunk 9]
```diff
@@ -94,6 +94,18 @@ FormulaError.REF = new FormulaError("#REF!");
  */
 FormulaError.VALUE = new FormulaError("#VALUE!");
 
+FormulaError.stringToError = (str) => {
+  const map = {
```
Comment (tyleradams) [inline formulas/error.js:98]: make sure it's in the map!

File: test/formulas/index.js [hunk 10]
```diff
@@ -1,11 +1,39 @@
...
+                const innerArr = [];
+                for (let col = ref.from.col - 1; col < ref.to.col; col++) {
+                    innerArr.push(data[row][col])
+                }
+                arr.push(innerArr);
+            }
+            return arr;
+        },
+        onVariable: name => {
+            return json[testName]["ON_VARIABLES"][name]
```
Comment (tyleradams) [inline test/formulas/index.js:31]: name it variables and keep it lower_snake_case

**Comment by marhong2** [comment]
const funs = Object.keys(TestCase);
        funs.forEach(fun => {
            const formulas = Object.keys(TestCase[fun]);
            formulas.forEach(formula => {
                let test = {}
                let expected = TestCase[fun][formula];
                if(expected.name){
                    expected = expected.name   
                }
                test["sheet"] = sheet
                test["expectedValue"] = expected 
                test["formula"] = formula
                json[formula] = test
            })
        })
        fs.appendFile('tests.json', JSON.stringify(json), (err) => {

File: formulas/error.js [hunk 11]
```diff
@@ -94,6 +94,20 @@ FormulaError.REF = new FormulaError("#REF!");
...
+    "#DIV/0!": FormulaError.DIV0,
+    "#N/A": FormulaError.NA,
+    "#NAME?": FormulaError.NAME,
+    "#NULL!":FormulaError.NULL, 
+    "#NUM!": FormulaError.NUM,
+    "#REF!": FormulaError.REF,
+    "#VALUE!": FormulaError.VALUE
+  }
+  if(map[str] != null)
```
Comment (tyleradams) [inline formulas/error.js:107]: if(str in map)

File: formulas/error.js [hunk 12]
```diff
@@ -94,6 +94,20 @@ FormulaError.REF = new FormulaError("#REF!");
...
+    return map[str]
+  return 0
```
Comment (tyleradams) [inline formulas/error.js:109]: Why 0

**Review by tyleradams — COMMENTED** [review]
Why is yarn lock changing?

**Comment by marhong2** [comment]
> 

I import a new library (Sinon) If this is not enough to explain the changes I can remake the code

File: test/formulas/index.js [hunk 13]
```diff
@@ -1,11 +1,57 @@
...
+        formula = json[testName]["formula"],
+        expected = json[testName]["expectedValue"]
+        clock =  sinon.useFakeTimers(date.getTime());
+    
+    if(data){
+        data.forEach((row, rowIndex) => row.forEach((value, colIndex) => {
+            if(value === "undefined")
+                data[rowIndex][colIndex] = undefined
+            if(FormulaError.stringToError(value))
+                data[rowIndex][colIndex] = FormulaError.stringToError(value);
```
Comment (jmorganmartin) [inline test/formulas/index.js:19]: Looking at how `stringToError()` is used, you might be able to remove it entirely and instead just check `FormulaError.errorMap.has(value)` and/or `FormulaError.errorMap.get(value)`

`FormulaError.errorMap` stores all the "generic" errors that are checked in `stringToError()`.

**Comment by jmorganmartin** [comment]
> > 
> 
> I import a new library (Sinon) If this is not enough to explain the changes I can remake the code (create new branch -> copy old code -> check to see if the changes to yarn.lock are the same)

`jscodeshift` is also added on this PR. It has dependencies on a bunch of newer versions of babel stuff, which constitute the majority of the changes to `yarn.lock`.

File: test/formulas/index.js [hunk 14]
```diff
@@ -1,11 +1,57 @@
...
+    let data = json[testName]["sheet"],
+        formula = json[testName]["formula"],
+        expected = json[testName]["expectedValue"]
+        clock =  sinon.useFakeTimers(date.getTime());
+    
+    if(data){
+        data.forEach((row, rowIndex) => row.forEach((value, colIndex) => {
+            if(value === "undefined")
+                data[rowIndex][colIndex] = undefined
+            if(FormulaError.errorMap.has(value))
```
Comment (tyleradams) [inline test/formulas/index.js:18]: Nice!

File: test/formulas/index.js [hunk 15]
```diff
@@ -1,11 +1,57 @@
...
+            }else if (result.error)
+                // For FormulaError
+                 expect(expected === result.error, `${formula} should equal ${expected.error}\n`).to.equal(true);
+            else
+                expect(result, `${formula} should equal ${expected}\n`).to.equal(expected);
+        })
+    });
+}
+
+Object.keys(json).map((testName) => test(testName))
```
Comment (tyleradams) [inline test/formulas/index.js:57]: You should guard this so it doesn't run by default. idk if you want this or something else as I'm unfamiliar with this part of jest.
https://stackoverflow.com/questions/4981891/node-js-equivalent-of-pythons-if-name-main

File: test/formulas/index.js [hunk 16]
```diff
@@ -1,11 +1,58 @@
...
+        },
+    });
+
+    describe(testName, function () {
+        it(formula, () => {  
+            let result = parser.parse(formula, {row:1, col:1});
+            if (typeof result === "number" && typeof expected === "number") {
+                expect(result, `${formula} should equal ${expected}\n`).to.closeTo(expected, 0.00000001);
+            }else if (result.error)
+                // For FormulaError
```
Comment (tyleradams) [inline test/formulas/index.js:49]: Remove comment

**Review by tyleradams — APPROVED** [review]

**Pull request merged**